### ISSUE TITLE
Possible boost race conditions

### ISSUE THREAD

**Issue by eschoeller** [Cacti/cacti#1608]
I may have a unique scenario here, so feel free to immediately dismiss this inquiry. 

I am intermittently getting the following types of errors while poller_boost.php is running:

<pre>
2018/05/04 19:28:49 - DBCALL ERROR: SQL Assoc Failed!, Error:1146, SQL:' (SELECT local_data_id, UNIX_TIMESTAMP(time) AS timestamp, rrd_name, output   FROM poller_output_boost   WHERE local_data_id = 16514   AND time < FROM_UNIXTIME(1525483729)) ORDER BY timestamp ASC, rrd_name ASC '
2018/05/04 19:28:49 - DBCALL ERROR: SQL Assoc Failed!, Error: Table 'cacti.poller_output_boost' doesn't exist
2018/05/04 19:28:49 - CMDPHP SQL Backtrace: (/scripts/flush_ds.php: 128 flush_ds)(/scripts/flush_ds.php: 152 boost_fetch_cache_check)(/lib/boost.php: 243 boost_process_poller_output)(/lib/boost.php: 681 db_fetch_assoc)(/lib/database.php: 360 db_fetch_assoc_prepared)(/lib/database.php: 401 cacti_debug_backtrace)
2018/05/04 19:46:58 - CMDPHP SQL Backtrace: (/graph_image-noauth.php: 104 rrdtool_function_graph)(/lib/rrd.php: 1148 boost_graph_cache_check)(/lib/boost.php: 321 boost_process_poller_output)(/lib/boost.php: 630 db_fetch_cell)(/lib/database.php: 233 db_fetch_cell_prepared)(/lib/database.php: 280 cacti_debug_backtrace)
</pre>

flush_ds.php is a script I have written which calls boost_fetch_cache_check for several data source IDs which I need to have updated on-disk in real-time. When boost_fetch_cache_check is executed while the poller_output_boost table has zero rows, it generates the "table doesn't exist" error. 

Similarly, graph_image-noatuh.php is a script which calls rrdtool_function_graph() which in turn would make a call to boost_graph_cache_check, which has a similar problem when the poller_output_boost table is empty. 

I've been able to mitigate this somewhat in my flush_ds script by executing the following SQL command:
<pre>'select table_rows from information_schema.TABLES where table_name = "poller_output_boost" and table_schema="cacti";'</pre> 
before each call to boost_fetch_cache_check to verify that the poller_output_boost table has greater than 0 rows. I still miss the timing by just enough to still generate the errors above. 
I'm not sure if there are situations elsewhere within Cacti itself which would trigger these situations, or if it it's just me.

**Comment by netniV** [comment]
The error given is not that †he table is empty, more that it doesn't exist at all.  Nothing should be removing the table as it's expected to be defined.

**Comment by eschoeller** [comment]
I agree. And the table *is* there, it's just empty, since boost is updating. But for some reason boost_fetch_cache_check throws a "does not exist" error instead.

**Comment by netniV** [comment]
This sounds more like a specific mysql issue rather than a cacti one.  There is nothing wrong with the queries.  Part of me wants to suggest restart the sql server and maybe even removing the log files so that things are clean but I'm no expert on mysql that would just be what I would do on my own stuff.

**Comment by eschoeller** [comment]
I have restarted the SQL server no less than 10 times this weekend while I've been working on this, but happy to do it again. So far today this is what's been logged for this issue:
<pre>
2018/05/06 03:53:15 - CMDPHP ERROR: A DB Exec Failed!, Error: Table 'cacti.poller_output_boost' doesn't exist
2018/05/06 05:54:16 - DBCALL ERROR: SQL Assoc Failed!, Error: Table 'cacti.poller_output_boost' doesn't exist
2018/05/06 15:13:16 - DBCALL ERROR: SQL Assoc Failed!, Error: Table 'cacti.poller_output_boost' doesn't exist
2018/05/06 15:46:56 - CMDPHP ERROR: A DB Exec Failed!, Error: Table 'cacti.poller_output_boost_arch_1525643181' doesn't exist
2018/05/06 15:46:56 - CMDPHP ERROR: A DB Exec Failed!, Error: Table 'cacti.poller_output_boost_arch_1525643181' doesn't exist
</pre>

**Comment by cigamit** [comment]
This was a re-write some time ago by another author than the original one.  It introduced the archive table, so there are periods of time that the boost table is unavailable.  This should have been done in a transaction to avoid such an issue.  In fact, if we move the boost table to InnoDB, we can force a transaction and remove these warnings.

**Comment by cigamit** [comment]
After doing some research, I had to make one more change related to transaction support.
